server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java

```diff
@@ -648,6 +650,13 @@ public IndexSearcher searcher() {
         return searcher;
     }
 
+    public FieldInfos getFieldInfos() {
+        if (this.fieldInfos == null) {
+            this.fieldInfos = searcher == null ? FieldInfos.EMPTY : FieldInfos.getMergedFieldInfos(searcher.getIndexReader());
+        }
+        return this.fieldInfos;
+    }
```

Comment:
`FieldInfos.getMergedFieldInfos` is not super expensive but it's not super cheap either since it runs in linear time with the number of fields x the number of segments. Would it be better to have a method called `fieldExists` there too that iterates over all segments and returns true if any of them has the field?

Reply:
++, have updated to iterate over the leaves.